### PR TITLE
Upgrade to Rails 5

### PR DESCRIPTION
**Pull request by allenwq**
See commit messages for the details

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -145,8 +145,8 @@
         get 'leaderboard' => 'leaderboard_settings#edit'
         patch 'leaderboard' => 'leaderboard_settings#update'
 
-        get 'topics' => 'discussion/topic_settings#edit', path: 'comments'
-        patch 'topics' => 'discussion/topic_settings#update', path: 'comments'
+        get 'topics' => 'discussion/topic_settings#edit'
+        patch 'topics' => 'discussion/topic_settings#update'
```
Comment (jeremyyap) [inline config/routes.rb]: @allenwq I didn't manage to fix this properly

Comment (allenwq) [reply config/routes.rb]: 👍  thanks for the info, will double check on this

File: lib/extensions/attachable/active_record/base.rb [hunk 2]
```diff
@@ -160,32 +159,55 @@ def attachment_reference
...
-        attribute_will_change!(ATTACHMENT_ATTRIBUTE)
+        mark_attachment_as_changed(attachment_reference)
         attachment_references.clear
       end
     end
+
+    # Tracks the the attachment_reference changes and support clear/revert the changes.
+    # `attachment_reference` is a virtual attribute, and changes tracking to such attributes are not well supported in
+    # rails 5: https://github.com/rails/rails/issues/25787
+    def attachment_reference_changed?
```
Comment (weiqingtoh) [inline lib/extensions/attachable/active_record/base.rb]: Once we upgrade to Rails 5.1 we should revert this commit?

Comment (allenwq) [reply lib/extensions/attachable/active_record/base.rb]: Possible, this commit gets merged in 5.1 which could support that (again): https://github.com/marketplacer/rails/commit/45e098898aa92189dfe98b93aee597602c7fc322 

But have to double check that time.

**Comment by allenwq** [comment]
The PR is ready for reviewing, I will take out some Rails 4 compatible commits and PR those separately.

**Pull request merged**